### PR TITLE
feat: treemap / sunburst / pie / donut charts

### PR DESCRIPTION
**Pull request by monfera**
## Summary

1st round of hierarchical, partitioning visualizations. Fixes https://github.com/elastic/elastic-charts/issues/84 including https://github.com/elastic/elastic-charts/issues/381#issuecomment-545179082
Other related issues: https://github.com/elastic/elastic-charts/issues/381 https://github.com/elastic/elastic-charts/issues/464 https://github.com/elastic/elastic-charts/issues/492

**Treemap:**

image

**Sunburst:**
image

**Donut:**
image

**Mid-transition from possible future tweening:**
image

Features we can lift:
- multiline, adaptive size, flexible shape text rendering
- text color inversion on light surfaces for accessibility
- hierarchical data processing with Typescript-only code (analogy: `GROUP BY ROLLUP`)
- compact, sunburst and treemap layout code in TS
- possibility of automatically generating configuration knobs
- many config options can be [tweened/animated](https://github.com/elastic/circline#circline) in the future

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [x] Any consumer-facing exports were added to `src/index.ts` (and stories only import from `../src` except for test data & storybook)
- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [x] Proper documentation or storybook story was added for features that require explanation or tutorials
- [ ] Unit tests were updated or added to match the most common scenarios
- [x] Each commit follows the [convention](https://github.com/elastic/elastic-charts/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by monfera** [comment]
Thanks @markov00 for the great notes and commits!

Point 1, first bullet, using index by the  `groupbyRollup` accessor, response commit https://github.com/elastic/elastic-charts/pull/493/commits/25e333e18bda5f38db0ef02d4609236a48abde23 notes: while a simple array of numbers of data like this was already possible: 

```tsx
<Partition
  id={'piechart'}
  data={[10, 20, 30]}
  valueFormatter={(d) => `${d}%`}
/>
```

the `layers: [layer0]` also needs to be specified when `nodeLabel` or `fillLabel` is not the default.  But `groupByRollup` is mandatory, and its default was the identity function, which was not good because it grouped equal numbers when specifying a number array - the commit changed that to the index. A better value would've been something like `(d: Datum) => uuid.v1()` or similar, but based on your suggestion, `(d: Datum, index: number) => index` is better because a user can check for range in the `groupByFunction` for multi-level partitions while `uuid` or similar is random. The use of an index is like an implicit `rowNumber` in an array/relation without actually having to store it.

With the commit, this now works, from the commit:
```tsx
<Partition
  id={'piechart'}
  data={[30, 20, 50]}
  valueFormatter={(d: Datum) => `${d}%`}
  layers={[
    {
      groupByRollup: (d: Datum, i: number) => (i < 2 ? 'A' : 'B'),
      nodeLabel: (d: Datum) => `Group ${d}`,
    },
    { 
      groupByRollup: (d: Datum, i: number) => i, 
      nodeLabel: (d: Datum) => `Item\u00A0${d}`
    },
  ]}
/>
```

Minor: https://github.com/elastic/elastic-charts/pull/493/commits/34262c41f548e5c611814d821d5217adf637cc37 was needed to make pies/treemaps show up; it could've been a more sophisticated workaround or an actual solution but you may have specific thoughts about how to _des-Cartes_ the check.

Minor: https://github.com/elastic/elastic-charts/pull/493/commits/286faca614f3c180a1103fad891406cdb836b3f4 and https://github.com/elastic/elastic-charts/pull/493/commits/25e333e18bda5f38db0ef02d4609236a48abde23  restored some former behaviors (see commit texts), please tell if the changes these revert were made for a specific reason. For example, the last two treemaps had superfluous texts showing up before https://github.com/elastic/elastic-charts/pull/493/commits/97c1b9d08f324b84ad6a3a6eb69858b47bb917ec

**Comment by monfera** [comment]
On the `nodeLabel` input:

> What if we pass in directly the datum? The user can return the appropriate label from the datum itself or from an external lookup table

It would work for single-layer partitioning.

For multi-layer partitioning, there's no single row in the input that'd correspond to the group node. So, in this case, the  closest thing  would be to pass on the array of direct constituents.

But then there's different input for the bottom layer vs. the higher layers. So, to make it general, the user should expect an array of items to make a determination  - it'd be an array of one for the bottom layer slices and an array of one or more for higher layers. It might be tedious for the user.

Another option is that, since a presumably unique value is coming out of the `groupByRollup` accessor, the  user joins on that inside the `nodeLabel` accessor, giving them a great deal of flexibility. I'll make a storybook example so we can test drive it

**Comment by monfera** [comment]
Marco I now see your reason for the change I restored in https://github.com/elastic/elastic-charts/commit/25e333e18bda5f38db0ef02d4609236a48abde23 - I'll need to figure out something and we should swap ideas on a call

**Comment by monfera** [comment]
Point 2, slice border taper-off effect - good call @markov00, the large slices look much better with no tapering. Empirically used a threshold of 15 degrees (see code comment in https://github.com/elastic/elastic-charts/pull/493/commits/6d141fc024aa75c557e7213bc66d8fe98b87d937)

**Comment by monfera** [comment]
Point 6 and 7 are done, the resulting paints are sometimes too dark and the text is often illegible (esp. on the darker hues) so  there's more to do, maybe through the `Theme`, let's discuss options

**Comment by monfera** [comment]
Point 9, multi-row outer labels: there are a few options there, each with their own strengths and weaknesses, let's catch up on this. Also, as it'd be a value-add relative to the current Kibana sunburst, it might not be a merge prerequisite, though we might need to add truncation which would be useful even once we add multi-row support

**Comment by monfera** [comment]
Point 8, splitting or not splitting `Partition`: what are the positives of splitting?  I can see one, which is that a couple of the options apply to only one or the other, eg. treemaps can't have outside linked labels (though they eventually should). 

The philosophical positives of keeping them together:
- the user intent is similar: show parts of a whole (in a single layer or hierarchically)
- though the "chart template" school of thought considers these charts separate, and they do use  differing shapes and areal partitioning, they're both fundamentally mapping a single metric for tabular, possibly rolled-up data to shape ink areas in some proportion
- there could be a couple of more versions in this group, eg. packed circles

The practical positives of keeping them  together:
- code and structure for config, data pre-processing, rendering etc. are almost fully shared
- the input data is of an identical structure, conducive to flat or nested partitioning
- the user can just toggle a control to switch to an (often) better representation eg. to treemap
- it'll be possible to apply a slick transition animation between sunburst, treemap and future forms

There may be things I overlook, or I may misunderstand the suggestion, so let's chat about it too

**Comment by monfera** [comment]
Point 4 on using `RecursivePartial` and `mergePartial` -  great request, thanks!

**Comment by monfera** [comment]
Comment on earlier feedback items for which no code commit was made: 
- for the time being, we discussed leaving the `<Partition>` together, ie. no separation to a Sunburst and a Treemap, due to near-identical config and semantics of the incoming data (we'll probably think about it some more in the future)
- multi-row outer labels: we need to circle back to it as it's not an easy addition, and it's not something the pie charts we replace do (ie. no regression in some important functionality)

File: stories/sunburst.tsx [hunk 1]
```diff
@@ -0,0 +1,699 @@
...
+        {
+          groupByRollup: (d: Datum) => d.sitc1,
+          nodeLabel: (d: Datum) => productLookup[d].name,
+          fillLabel: { textInvertible: true },
+          shape: {
+            fillColor: defaultFillColor(interpolatorCET2s),
+          },
+        },
+      ]}
+      config={{ ...config, hierarchicalLayout: PartitionLayouts.sunburst }}
```
Comment (markov00) [inline stories/sunburst.tsx]: there are a lot of  `...config`, so please remove them to show a cleaner example

File: stories/sunburst.tsx [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+        {
+          groupByRollup: (d: Datum) => d.sitc1,
+          nodeLabel: (d: Datum) => productLookup[d].name,
+          fillLabel: { textInvertible: true },
+          shape: {
+            fillColor: defaultFillColor(interpolatorTurbo),
+          },
+        },
+      ]}
+      config={{ ...config, hierarchicalLayout: PartitionLayouts.sunburst }}
```
[the remark this answers is not in the document]
Comment (markov00) [reply stories/sunburst.tsx]: we can also remove the `hierarchicalLayout` if the sunburst is the default type. I think if we provide the most simplest and cleanest example for the first chart, the better is for the user

File: package.json [hunk 3]
```diff
@@ -136,7 +137,8 @@
     "ts-jest": "^24.1.0",
     "ts-loader": "^6.2.1",
     "ts-prune": "^0.3.0",
-    "typescript": "^3.7.0",
+    "typescript": "^3.7.2",
+    "utility-types": "^3.10.0",
```
[the remark this answers is not in the document]
Comment (monfera) [reply package.json]: 👍

File: src/chart_types/partition_chart/layout/circline_geometry.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
+import { CirclArc, Circline, CirclinePredicate, Distance, PointObject, RingSector } from './types/geometry_types';
+import { tau } from './utils/math';
+
+function euclideanDistance({ x: X, y: Y }: PointObject, { x, y }: PointObject): Distance {
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/circline_geometry.ts]: 👍

File: src/chart_types/partition_chart/layout/circline_geometry.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+function fullyContained(C: Circline, c: Circline): boolean {
+  return euclideanDistance(C, c) + c.r <= C.r;
+}
+function noOverlap(C: Circline, c: Circline): boolean {
+  return euclideanDistance(C, c) >= C.r + c.r;
+}
+
+function circlineIntersect(c1: Circline, c2: Circline): PointObject[] {
+  const D = Math.sqrt((c1.x - c2.x) * (c1.x - c2.x) + (c1.y - c2.y) * (c1.y - c2.y));
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/circline_geometry.ts]: 👍

File: src/chart_types/partition_chart/layout/config/config.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
+import { palettes } from '../../../../mocks/hierarchical/palettes';
+import { Config, PartitionLayouts, Numeric } from '../types/config_types';
+import { goldenRatio, tau } from '../utils/math';
+
+// todo add more structure to the config (like `linkLabel`, which is already nested)
+// todo add more configurability, eg. font weight
+// todo allow an option for direct coloring of the slices if the caller desires so
+// todo separate the concept of real min/max values from the concept of min/max range for playground style randomization
+// todo abstract out types for tuple-like things (eg. font descriptions have family, size, style...)
+// todo use preexisting types for describing fonts, paint styles etc.
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/config/config.ts]: 👍

[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/config/config.ts]: Thanks, as many of these have been addressed, I'll make a new one for the current list of improvement opportunities

File: src/chart_types/partition_chart/layout/utils/treemap.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
+import { ArrayEntry, childrenKey, entryValue, HierarchyOfArrays } from './group_by_rollup';
+import { Part } from '../types/types';
+import { goldenRatio } from './math';
+
+const maxPaddingRatio = 0.0256197; // this limits area distortion to <10% (which occurs due to pixel padding) with very small rectangles
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/treemap.ts]: 👍

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (monfera) [reply package.json]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/4df9dfb677e8481345efd4f5760f59ccf3f37990

File: src/chart_types/partition_chart/layout/circline_geometry.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/circline_geometry.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/8e1be0535127218c014db4d8da17dc5b22733b60

File: src/chart_types/partition_chart/layout/circline_geometry.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/circline_geometry.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/8e1be0535127218c014db4d8da17dc5b22733b60

File: src/chart_types/partition_chart/layout/utils/math.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import { wrapToTau } from '../geometry';
+
+export const rightAngle = Math.PI / 2;
+export const goldenRatio = 1.618;
+export const tau = 2 * Math.PI;
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/math.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/8e1be0535127218c014db4d8da17dc5b22733b60

File: src/chart_types/partition_chart/layout/utils/treemap.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/treemap.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/8e1be0535127218c014db4d8da17dc5b22733b60

File: src/chart_types/partition_chart/layout/utils/treemap.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    .concat(
+      ...vector.map(({ node, x0, y0, x1, y1 }) => {
+        const childrenNodes = entryValue(node)[childrenKey];
+        const width = x1 - x0;
+        const height = y1 - y0;
+        const padding = Math.min(paddingAccessor(node), width * maxPaddingRatio * 2, height * maxPaddingRatio * 2);
+        const innerWidth = width - 2 * padding;
+        const innerHeight = height - 2 * padding;
+        return childrenNodes && childrenNodes.length
+          ? treemap(
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/treemap.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/3b9a8bdf9b8e845df866c9315c398496b9395964

File: src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+function ringSectorInnerRadius(innerRadius: Radian, ringThickness: Distance) {
+  return (d: ShapeTreeNode): Radius => innerRadius + (d.y0 as number) * ringThickness;
+}
+function ringSectorOuterRadius(innerRadius: Radian, ringThickness: Distance) {
+  return (d: ShapeTreeNode): Radius => innerRadius + ((d.y0 as number) + 1) * ringThickness;
+}
+
+const infinityRadius = 1e4; // far enough for a sub-2px precision on a 4k screen, good enough for text bounds; 64 bit floats still work well with it
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/9f6226d3ad1cfaf183b751cd5119961121a9a973

File: src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+      fillRectangleWidth,
+      fillRectangleHeight,
+    } = config;
+    const r =
+      (fillOutside ? ringSectorOuterRadius : ringSectorInnerRadius)(innerRadius, ringThickness)(ringSector) +
+      circlePadding * 2;
+    const R = Math.max(
+      r,
+      ringSectorOuterRadius(innerRadius, ringThickness)(ringSector) - circlePadding + (fillOutside ? radiusOutside : 0),
+    );
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/8873a5e20b9d032d8aa8948823a0bca107ec81e2

File: src/chart_types/partition_chart/layout/types/geometry_types.ts [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+  x: Coordinate = NaN;
+  y: Coordinate = NaN;
+  r: Radius = NaN;
+}
+
+export interface CirclinePredicate extends Circline {
+  inside: boolean;
+}
+
+export interface CirclArc extends Circline {
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/types/geometry_types.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/8873a5e20b9d032d8aa8948823a0bca107ec81e2

File: src/chart_types/partition_chart/layout/types/types.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+export type TextMeasure = (font: string, texts: string[]) => TextMetrics[];
+
+// Part-to-whole visualizations such as treemap, sunburst, pie hinge on an aggregation
+// function such that the value is independent of the order of how the constituents are aggregated
+// https://en.wikipedia.org/wiki/Associative_property
+// Hierarchical, space-filling part-to-whole visualizations also need that the
+// the value of a node is equal to the sum of the values of its children
+// https://mboehm7.github.io/teaching/ss19_dbs/04_RelationalAlgebra.pdf p21
+// It's now `count` and `sum` but subject to change
+export type AdditiveAggregation = 'count' | 'sum';
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/types/types.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/ec20035ee8d8089d3421e463a074a352484be2b1

File: src/chart_types/partition_chart/layout/utils/group_by_rollup.ts [hunk 14]
```diff
@@ -0,0 +1,170 @@
+import { Relation } from '../types/types';
+
+export const aggregateKey = 'value'; // todo later switch back to 'aggregate'
+export const depthKey = 'depth';
+export const childrenKey = 'children';
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/group_by_rollup.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/9f6226d3ad1cfaf183b751cd5119961121a9a973

File: src/chart_types/partition_chart/renderer/canvas/canvas_renderers.ts [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+  RowSet,
+  ShapeViewModel,
+  TextRow,
+} from '../../layout/types/viewmodel_types';
+import { tau } from '../../layout/utils/math';
+import { PartitionLayouts } from '../../layout/types/config_types';
+
+// the burnout avoidance in the center of the pie
+const lineWidthMult = 10; // border can be a maximum 1/lineWidthMult - th of the sector angle, otherwise the border would dominate
+const taperOffLimit = 50; // taper off within a radius of taperOffLimit to avoid burnout in the middle of the pie when there are hundreds of pies
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/renderer/canvas/canvas_renderers.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/9f6226d3ad1cfaf183b751cd5119961121a9a973

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+      // @ts-ignore
+      const s = source[key];
+      // @ts-ignore
+      const t = target[key];
+      // @ts-ignore
+      target[key] = t && s && typeof t === 'object' && typeof s === 'object' ? objectAssign(t, s) : s;
+    });
+  });
+  return target;
+}
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/calcs.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/d23b3596cd0753d7d004fd67f6335ac1259ca01a

File: src/chart_types/partition_chart/layout/utils/calcs.ts [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+export function deepTween(target: object, source: object, ratio: Ratio) {
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/calcs.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/d23b3596cd0753d7d004fd67f6335ac1259ca01a

File: src/chart_types/partition_chart/layout/utils/d3_utils.ts [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+  return d3Rgb(cssColorSpecifier) || { r: 255, g: 0, b: 0, opacity: 1 };
+}
+
+function argsToRGB(r: number, g: number, b: number, opacity: number): RgbObject {
+  return d3Rgb(r, g, b, opacity) || { r: 255, g: 0, b: 0, opacity: 1 };
+}
+
+export function argsToRGBString(r: number, g: number, b: number, opacity: number): string {
+  // d3.rgb returns an Rgb instance, which has a specialized `toString` method
+  return argsToRGB(r, g, b, opacity).toString();
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/d3_utils.ts]: Thanks for the catch, I did it like the above while fixing the function type too (we expect to have meaningful `.toString()` on the result of the `argsToRGB` call) and some DRY

commit: https://github.com/elastic/elastic-charts/pull/493/commits/9b6e25840d96792195b16714f93a84dfdb7b9e42

File: src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+  cy: Coordinate,
+  radialOffset: Distance,
+  rotation: Radian,
+  fontSize: number,
+  offsetSign: -1 | 0 | 1,
+) {
+  const rowCentroidY = cy;
+  const r = infinityRadius;
+  const offset = (offsetSign * fontSize) / 2;
+  const anotherOffset = rowCentroidY - cy + offset;
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/d23b3596cd0753d7d004fd67f6335ac1259ca01a

File: src/chart_types/partition_chart/layout/types/config_types.ts [hunk 20]
```diff
@@ -0,0 +1,126 @@
+import { Distance, Pixels, Radian, Radius, Ratio, SizeRatio, TimeMs } from './geometry_types';
+import { Color, FontWeight } from './types';
+import { $Values as Values } from 'utility-types';
+
+export const PartitionLayouts = Object.freeze({
+  sunburst: 'sunburst',
+  treemap: 'treemap',
+});
+
+export type HierarchicalLayout = Values<typeof PartitionLayouts>; // could use ValuesType<typeof HierarchicalChartTypes>
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/types/config_types.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/e5dc415f7afbac32b680e47dba915a16b53d8151

File: src/chart_types/partition_chart/layout/utils/group_by_rollup.ts [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+      const node = keyExists && pointer.get(key);
+      const childrenMap = node ? node[childrenKey] : new Map();
+      const aggregate = node ? node[aggregateKey] : identity();
+      const reductionValue = reducer(aggregate, valueAccessor(n));
+      pointer.set(key, {
+        [aggregateKey]: reductionValue,
+        [depthKey]: i,
+        ...(!last && { [childrenKey]: childrenMap }),
+      });
+      if (childrenMap) {
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/utils/group_by_rollup.ts]: Marco, if you mean something like
```
    keyAccessors.forEach((keyAccessor, i) => {
      const key = keyAccessor(n, index);
      const keyExists = pointer.has(key);
      const node = keyExists && pointer.get(key);
      const childrenMap = node ? node[CHILDREN_KEY] : new Map();
      if (childrenMap) {
        // will always be true except when exiting from forEach, ie. upon encountering the leaf node
        pointer = childrenMap;
        return; // with or without the return
      }
      const aggregate = node ? node[AGGREGATE_KEY] : identity();
      const reductionValue = reducer(aggregate, valueAccessor(n));
      const last = i === keyCount - 1;
      pointer.set(key, {
        [AGGREGATE_KEY]: reductionValue,
        [DEPTH_KEY]: i,
        ...(!last && { [CHILDREN_KEY]: childrenMap }),
      });
    });
```
its meaning wouldn't be the same, as the `Map` value for `key` should still be set, but I might misunderstand your suggestion

File: src/chart_types/partition_chart/state/selectors/scenegraph.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+  return (node: ShapeTreeNode) => {
+    const accessorFn = layers[node[depthKey] - 1].nodeLabel || identity;
+    return `${accessorFn(node.dataName)}`;
+  };
+}
+
+export function render(partitionSpec: PartitionSpec, parentDimensions: Dimensions, theme: Theme): ShapeViewModel {
+  const { width, height } = parentDimensions;
+  // eslint-disable-next-line no-console
+  console.log(theme);
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/state/selectors/scenegraph.ts]: Is it OK if we remove this line once the code takes colors from Theme? (I put it on the list in the new gh issue) I'm a bit uncertain at the moment as to what information we should take from the Theme, but if we remove this line, the `theme` will be an unused parameter and needs further unraveling in the caller, while it should instead be actually used

File: src/chart_types/partition_chart/state/chart_state.tsx [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+  }
+  getLegendItemsValues() {
+    return EMPTY_MAP;
+  }
+  chartRenderer() {
+    return (
+      <React.Fragment>
+        <Partition />
+      </React.Fragment>
+    );
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/state/chart_state.tsx]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/10047f6d4d28ce1eee9a28b30b71c246758938f9

File: stories/sunburst.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (monfera) [reply stories/sunburst.tsx]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/d5877ff10383bae5232c8e8d63c9e5d623d9d1f0

File: stories/sunburst.tsx [hunk 24]
```diff
@@ -0,0 +1,699 @@
...
+        {
+          groupByRollup: (d: Datum) => d.sitc1,
+          nodeLabel: (d: Datum) => productLookup[d].name,
+          fillLabel: { textInvertible: true },
+          shape: {
+            fillColor: defaultFillColor(interpolatorCET2s),
+          },
+        },
+      ]}
+      config={{ ...config, hierarchicalLayout: PartitionLayouts.sunburst }}
```
[the remark this answers is not in the document]
Comment (monfera) [reply stories/sunburst.tsx]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/d5877ff10383bae5232c8e8d63c9e5d623d9d1f0

File: stories/sunburst.tsx [hunk 25]
```diff
@@ -0,0 +1,699 @@
...
+        {
+          groupByRollup: (d: Datum) => d.sitc1,
+          nodeLabel: (d: Datum) => d,
+          fillLabel: { textInvertible: true },
+          shape: {
+            fillColor: defaultFillColor(interpolatorCET2s),
+          },
+        },
+      ]}
+      config={{ ...config, hierarchicalLayout: PartitionLayouts.sunburst }}
```
[the remark this answers is not in the document]
Comment (monfera) [reply stories/sunburst.tsx]: https://github.com/elastic/elastic-charts/pull/493/commits/d5877ff10383bae5232c8e8d63c9e5d623d9d1f0

File: stories/sunburst.tsx [hunk 26]
```diff
@@ -0,0 +1,699 @@
...
+        {
+          groupByRollup: (d: Datum) => d.sitc1,
+          nodeLabel: (d: Datum) => productLookup[d].name,
+          fillLabel: { textInvertible: true },
+          shape: {
+            fillColor: defaultFillColor(interpolatorCET2s),
+          },
+        },
+      ]}
+      config={{ ...config, hierarchicalLayout: PartitionLayouts.sunburst }}
```
[the remark this answers is not in the document]
Comment (monfera) [reply stories/sunburst.tsx]: https://github.com/elastic/elastic-charts/pull/493/commits/d5877ff10383bae5232c8e8d63c9e5d623d9d1f0

File: stories/sunburst.tsx [hunk 27]
```diff
@@ -0,0 +1,699 @@
...
```
[the remark this answers is not in the document]
Comment (monfera) [reply stories/sunburst.tsx]: https://github.com/elastic/elastic-charts/pull/493/commits/d5877ff10383bae5232c8e8d63c9e5d623d9d1f0

File: stories/sunburst.tsx (hunk 1 above)
Comment (monfera) [reply stories/sunburst.tsx]: https://github.com/elastic/elastic-charts/pull/493/commits/d5877ff10383bae5232c8e8d63c9e5d623d9d1f0

I removed the now unneeded spreads in substructures, and there was still one in `treemap.tsx`

File: stories/sunburst.tsx [hunk 28]
```diff
@@ -0,0 +1,699 @@
...
+// style calcs
+const interpolatorCET2s = hueInterpolator(palettes.CET2s.map(([r, g, b]) => [r, g, b, 0.8]));
+const interpolatorTurbo = hueInterpolator(palettes.turbo.map(([r, g, b]) => [r, g, b, 0.8]));
+
+const defaultFillColor = (colorMaker: any) => (d: any, i: number, a: any[]) => colorMaker(i / (a.length + 1));
+
+export const SimplePieChart = () => (
+  <Chart className={'story-chart'}>
+    <Partition
+      id={getSpecId('spec_' + getRandomNumber())}
```
[the remark this answers is not in the document]
Comment (monfera) [reply stories/sunburst.tsx]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/1e56e661d29531e210c320a1b1624098fae3aa21

**Comment by monfera** [comment]
Failed CI runs solved by checking in `yarn.lock`

File: src/chart_types/partition_chart/specs/index.ts [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+export interface PartitionSpec extends Spec {
+  specType: typeof SpecTypes.Series;
+  chartType: typeof ChartTypes.Partition;
+  config: RecursivePartial<Config>;
+  data: Datum[];
+  valueAccessor: AccessorFn;
+  valueFormatter: AccessorFn;
+  layers: Layer[];
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/specs/index.ts]: commit https://github.com/elastic/elastic-charts/pull/493/commits/2623d0f54c34f9b0c267e42d45eeca852f059771

File: stories/sunburst.tsx [hunk 30]
```diff
@@ -0,0 +1,699 @@
+import { Chart, getSpecId, Partition } from '../src';
+import { mocks } from '../src/mocks/hierarchical/index';
+import { config } from '../src/chart_types/partition_chart/layout/config/config';
+import { arrayToLookup, hueInterpolator } from '../src/chart_types/partition_chart/layout/utils/calcs';
+import { Datum } from '../src/chart_types/xy_chart/utils/specs';
+import { productDimension, regionDimension, countryDimension } from '../src/mocks/hierarchical/dimension_codes';
+import React from 'react';
+import { PartitionLayouts } from '../src/chart_types/partition_chart/layout/types/config_types';
+import { getRandomNumber } from '../src/mocks/utils';
+import { palettes } from '../src/mocks/hierarchical/palettes';
```
[the remark this answers is not in the document]
Comment (monfera) [reply stories/sunburst.tsx]: commit https://github.com/elastic/elastic-charts/pull/493/commits/2623d0f54c34f9b0c267e42d45eeca852f059771

[the remark this answers is not in the document]
Comment (monfera) [reply stories/sunburst.tsx]: also, commit https://github.com/elastic/elastic-charts/pull/493/commits/e8a6a9d8c49faebee48d1981c07a30663e846ffc merges some import lines

File: src/chart_types/partition_chart/layout/types/config_types.ts [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+
+export type EasingFunction = (x: Ratio) => Ratio;
+
+export interface AnimKeyframe {
+  time: number;
+  easingFunction: EasingFunction;
+  keyframeConfig: Partial<StaticConfig>;
+}
+
+export interface Config extends StaticConfig {
```
[the remark this answers is not in the document]
Comment (monfera) [reply src/chart_types/partition_chart/layout/types/config_types.ts]: commit: https://github.com/elastic/elastic-charts/pull/493/commits/2623d0f54c34f9b0c267e42d45eeca852f059771

**Pull request merged**